conda_libmamba_solver/shards.py

```diff
@@ -325,58 +341,106 @@ def repodata_shards(url, cache: RepodataCache) -> bytes:
+    # If we fall back to monolithic repodata.json, the standard fetch code will
+    # load the state again in text mode.
+    try:
+        # by using read_bytes, a possible UnicodeDecodeError should be converted
+        # to a JSONDecodeError. But a valid json that is not a dict will fail on
+        # .update().
+        repo_cache.state.update(json.loads(repo_cache.cache_path_state.read_bytes()))
+    except (FileNotFoundError, json.JSONDecodeError):
+        pass
+
+    cache_state = repo_cache.state
+
+    if cache is None:
+        cache = shards_cache.ShardCache(Path(conda.gateways.repodata.create_cache_dir()))
 
     if cache_state.should_check_format("shards"):
+        if not repo_cache.cache_path_shards.exists():
+            # avoid 304 not modified if we don't have the file
+            cache_state.etag = ""
+            cache_state.mod = ""
         try:
             # look for shards index
             shards_index_url = f"{sd.url_w_subdir}/repodata_shards.msgpack.zst"
-            found = repodata_shards(shards_index_url, cache)
+            found = repodata_shards(shards_index_url, repo_cache)
             cache_state.set_has_format("shards", True)
             # this will also set state["refresh_ns"] = time.time_ns(); we could
             # call cache.refresh() if we got a 304 instead:
-            cache.save(found)
+            repo_cache.save(found)
 
             # basic parse (move into caller?)
-            shards_index: ShardsIndex = msgpack.loads(zstandard.decompress(found))  # type: ignore
-            shards = Shards(
-                shards_index,
-                shards_index_url,
-                shards_cache.ShardCache(Path(conda.gateways.repodata.create_cache_dir())),
-            )
+            shards_index: ShardsIndexDict = msgpack.loads(zstandard.decompress(found))  # type: ignore
+            shards = Shards(shards_index, shards_index_url, cache)
             return shards
 
         except (HTTPError, conda.gateways.repodata.RepodataIsEmpty):
             # fetch repodata.json / repodata.json.zst instead
             cache_state.set_has_format("shards", False)
-            cache.refresh(refresh_ns=1)  # expired but not falsy
+            repo_cache.refresh()
 
     return None
+
+
+def batch_retrieve_from_cache(sharded: list[Shards], packages: list[str]):
+    """
+    Given a list of Shards objects and a list of package names, fetch all URLs
+    from a shared local cache, and update Shards with those per-package shards.
+    Return the remaining URLs that must be fetched from the network.
+    """
+    sharded = [shardlike for shardlike in sharded if isinstance(shardlike, Shards)]
+
+    wanted = []
+    for shard in sharded:
+        for package_name in packages:
+            if package_name in shard:
+                wanted.append((shard, package_name, shard.shard_url(package_name)))
+
+    log.debug("%d shards to fetch", len(wanted))
+
+    shared_shard_cache = sharded[0].shards_cache
```

Comment:
This will raise `IndexError` if no shards were found in `sharded`.

Reply:
Worked around, but we might still subset normal repodata. We'll see if we can afford to do that, or not, once the tempfiles are eliminated.